storage: acl: refactor

stephenplusplus opened the pull request:
Fixes #351 

This introduces support for a nicer API to ACL manipulations. The new methods are broken into groups representing a permission level, owners, readers, and writers. Each group has access methods to add or delete various entities: allAuthenticatedUsers, allUsers, domains, groups, projects, and users.

The full list of methods:

``` js
var acl = (myFile || myBucket).(acl || acl.default).(owners || readers || writers);

acl.addAllAuthenticatedUsers(function(err, aclObject) {});
acl.deleteAllAuthenticatedUsers(function(err) {});

acl.addAllUsers(function(err, aclObject) {});
acl.deleteAllUsers(function(err) {});

acl.addDomain(function(err, aclObject) {});
acl.deleteDomain(function(err) {});

acl.addGroup(function(err, aclObject) {});
acl.deleteGroup(function(err) {});

acl.addProject(function(err, aclObject) {});
acl.deleteProject(function(err) {});

acl.addUser(function(err, aclObject) {});
acl.deleteUser(function(err) {});
```

Additionally, a couple methods were added to simplify common operations; making a bucket or file publicly readable or the inverse.

``` js
(myFile || myBucket).makePrivate(function(err) {});
(myFile || myBucket).makePublic(function(err) {});
```

This will run `setMetadata` with a `predefinedAcl` value set to `private` or `publicRead`.

stephenplusplus commented on lib/storage/acl.js:
```diff
@@ -295,13 +418,55 @@ Acl.prototype.update = function(options, callback) {
 Acl.prototype.makeAclObject_ = function(accessControlObject) {
   var obj = {
-    scope: accessControlObject.scope,
+    scope: accessControlObject.entity,
```
Not sure if the API changed, this was always a bug, or I'm crazy. I noticed the API was responding with an `entity` property and not `scope`. I changed this to support the API.

ryanseys replied:
And the tests didn't catch it? :(

stephenplusplus replied:
I believe we had regression tests checking the part of the response that was consistent, `role`. Our new tests check for `scope` as well.

jgeewax replied:
So do we want to create a ticket to change ours from `scope` to `entity` ?

stephenplusplus replied:
We could do that with the benefit of being consistent with the upstream API, but I think we can still use `scope` if we think the term "scope" is clearer to the user. I'm happy with either option.

ryanseys replied:
The API never changed, it was always `entity` ([see here](https://github.com/google/google-api-nodejs-client/tree/master/apis/storage)) so I'm not sure where `scope` came from, may have been a simplification of the terminology which makes sense to me. For upstream support it makes sense to change to `entity` now that we have more reasonable methods to interacting with ACLs, people who want to do it manually can stick to the docs more closely.

jgeewax replied:
I'm +1 on entity then.

stephenplusplus commented on lib/storage/acl.js:
```diff
@@ -295,13 +418,55 @@ Acl.prototype.update = function(options, callback) {
+Acl.prototype.makePrivate = function(callback) {
+  this.delete({
+    scope: 'allUsers'
+  }, callback);
```
:us: Flagging this for attention. Not sure this is the intended effect making a bucket/object private is meant to have.

ryanseys replied:
This one seems reasonable for files. But if it's a bucket, again, might need to do defaults as well.

stephenplusplus commented on lib/storage/acl.js:
```diff
@@ -295,13 +418,55 @@ Acl.prototype.update = function(options, callback) {
+    scope: 'allUsers',
+    role: 'READER'
+  }, callback);
+};
```
(same as above)

:us: Flagging this for attention. Not sure this is the intended effect making a bucket/object public is meant to have.

ryanseys replied:
Hmm... yeah if this is a bucket, should it be `acl.default.add()`? go go gadget @jgeewax! (because you suggested the methods)

stephenplusplus replied:
`makePublic` can be done on a bucket via `myBucket.acl.makePublic()` and also on the default object as well: `myBucket.acl.default.makePublic()`.

ryanseys replied:
What about:

``` js
if(this.isFile()) {
  this.acl.add({
    scope: 'allUsers',
    role: 'READER'
  }, callback);
} 
else if(this.isBucket()) {
  this.acl.add({
   scope: 'allUsers',
   role: 'READER'
 }, function(err, resp) {
    if(!err) {
      this.acl.default.add({
        scope: 'allUsers',
        role: 'READER'
      }, callback);
    } else {
      // uh oh
    }
  });
}
else {
  // nope chuck testa
}

stephenplusplus replied:
Basically, if it's a file, make it public. If it's a bucket, make the default acl public as well as the current files in the bucket?

I could see that being confusing since we have the `myBucket.acl.default` object as well, which would have its own `makePublic`. And in the case where a user wants to only make their current OR future files public, they would have to do the long-hand method.

ryanseys replied:
Yep just realized that and came back to say that we should handle both separately. :+1:

ryanseys commented:
hmm... do you like `delete` or `remove` more? I'm feeling `remove` more because `acl.writers.removeUser` sounds more realistic than `acl.writers.deleteUser` but because we already have `delete` it might make sense to keep it that way... tough call.

stephenplusplus commented:
I prefer `delete` for consistency across our library. Files, buckets, datasets, etc. are all deleted, and as you pointed out, we have `acl.delete` as well.

ryanseys commented:
Works for me! I've been writing too much ruby recently.

stephenplusplus commented:
ha_ha()

ryanseys commented:
> `ha_ha()`

No brackets necessary.

stephenplusplus commented:
I think we have PR review inception here.

ryanseys commented on lib/storage/acl.js:
```diff
@@ -328,3 +493,90 @@ Acl.prototype.makeReq_ = function(method, path, query, body, callback) {
+
+        that[accessMethod]({
+          scope: scope,
+          role: role
```
No requirement to be uppercase e.g. OWNER, READER, WRITER?

stephenplusplus replied:
Judging from the API working in the regression tests, it's not required to be uppercase. Nonetheless, it's an oversight that I didn't. Fixed: https://github.com/GoogleCloudPlatform/gcloud-node/commit/aefdd45899ee4e6ec2bd012ff3e195bf9ebbdcf3

ryanseys commented on regression/storage.js:
```diff
@@ -149,6 +149,18 @@ describe('storage', function() {
+      });
+
+      it('should make a bucket private', function(done) {
+        bucket.acl.makePrivate(done);
```
assert role and scope like above test.

ryanseys replied:
Oops this be regression tests.

ryanseys replied:
Wait, how do we know this will be executed in order? Shouldn't this be done inside the other test?

stephenplusplus replied:
Deleting comes back with an empty response body.

stephenplusplus replied:
Updating any docs that didn't list the arguments to the callback correctly.

stephenplusplus replied:
Tests merged: https://github.com/stephenplusplus/gcloud-node/commit/3a70ef724dcd18b19ee083225cea9809d61f342d

ryanseys commented:
Looks good to me :)

ryanseys commented on lib/storage/acl.js:
```diff
@@ -295,13 +418,55 @@ Acl.prototype.update = function(options, callback) {
+ * @alias acl.makePublic
+ *
+ * @example
+ * myBucket.acl.makePublic(function(err, aclObject) {});
```
`aclObject` --> `acl` ?

stephenplusplus replied:
The examples have used `aclObject` since we added acl support. I tried to make the name overly descriptive just to help. Do you think we should switch them all to `acl`?

ryanseys replied:
Hmm... if you are keeping them `aclObject` then can you change [this](https://github.com/stephenplusplus/gcloud-node/commit/6db8b3a64ebefd7cc91ad3870568e0d2dfcbba44#diff-712237668475919bc2aceb8288701749R508) to `aclObject` ?

stephenplusplus replied:
I used `acl` for short-hand, as using `aclObject` would force a new line. That's our private, developer-documentation. I don't think we'll be confused :)

ryanseys replied:
Oh, well I was confused.

stephenplusplus replied:
What confused you, that it was private documentation? It's marked `@private` to keep it out of the published docs on the site. I could add the `/*! Developer Documentation` heading, though that's typically for broader documentation purposes, such as "The methods that follow are for...". This is just the documentation for this class.

If the confusion was `acl` vs `aclObject`, were you confused about the object itself, or that `acl` in this example wouldn't be the same object as `aclObject` from another example?

ryanseys replied:
I just wasn't sure about the consistency of the variable names. The fact that the 80-char line limit trumps code consistency upsets me but I'm over it. I didn't notice the `@private` so I'll be more careful about that next time.

ryanseys commented on lib/storage/acl.js:
```diff
@@ -67,11 +69,132 @@ var util = require('../common/util.js');
+ * myFile.acl.add({
+ *   scope: 'user-email@example.com',
+ *   role: storage.acl.WRITER_ROLE
+ * }, function(err, aclObject) {});
```
`aclObject` --> `acl`?

ryanseys commented:
I think this is acceptable for merge. Should probably wait for sign-off by @jgeewax for the makePrivate and makePublic however.

stephenplusplus commented:
Updated `makePublic()` and `makePrivate()`: https://github.com/stephenplusplus/gcloud-node/commit/fd35f2746b1620ca180961be811ca6d2c520fc10

They are now on the File and Bucket object themselves.

ryanseys commented:
Is this ready for another look?

stephenplusplus commented:
Yes, and I should have explained the new `makePublic()` and `makePrivate()` methods use a shorthand "predefinedAcl" value that Ryan found. When a user makes an object public or private, it updates the "predefinedAcl" value as a metadata request to `publicRead` or `private`.

jgeewax commented on test/storage/acl.js:
```diff
@@ -20,7 +20,8 @@
 var assert = require('assert');
-var storage = require('../../lib/storage/index.js');
+var async = require('async');
+var Storage = require('../../lib/storage/index.js');
```
Why'd we change this to uppercase Storage? Just curious.

ryanseys replied:
Capitalization suggests it's not an instance, which it isn't.

jgeewax replied:
Gotcha - thanks.

jgeewax commented on lib/storage/bucket.js:
```diff
@@ -288,8 +306,51 @@ Bucket.prototype.getMetadata = function(callback) {
+  callback = callback || util.noop;
+
+  this.setMetadata({
+    predefinedAcl: 'publicRead'
```
Does this make all the files in the bucket public as well? ie,

- create bucket
- create file in bucket
- `file.makePrivate()`
- `bucket.makePublic()`

Is file public? (I argue it should be.)

ryanseys replied:
From the [docs](https://cloud.google.com/storage/docs/accesscontrol#predefined-acl):

> Gives the bucket or object owner OWNER permission and gives all users, both authenticated and anonymous, READER permission. When you apply this to an object, anyone on the Internet can read the object without authenticating. When you apply this to a bucket, anyone on the Internet can list objects without authenticating. 

If by "list objects", they mean read the objects contents, then yes.

ryanseys replied:
Should we consider using `Cache-Control:private, max-age=0, no-transform` on the request to make the changes propagate immediately?

stephenplusplus replied:
I tested @jgeewax's series of events, and the file remained private after making the bucket public. I'm not sure the URL I would visit to see a directory listing of my bucket to confirm whether it is included in that list.

jgeewax replied:
+1 to making changes propagate immediately.

Re "list objects" : I don't believe that it means objects are readable... List means GET /bucket will tell you the list of keys, but GET /bucket/objects/object is what we're looking for.

I think `bucket.makePublic` needs to do more than just mark the bucket as public. It should also (depending on arguments):

- set the defaultACL to 'public' (so all future items are made public) (ie, `if (future) { /* set default ACL to public */ }`)
- set all existing items to 'public' (so all existing items are public (ie `if (recursive) { /* set everything to public */ }`)

ryanseys replied:
> I'm not sure the URL I would visit to see a directory listing of my bucket to confirm whether it is included in that list.

http://storage.googleapis.com/my-bucket-name/ should work.

ryanseys commented:
@stephenplusplus For now, you should just remove `makePublic` and `makePrivate` and we can land these changes. Those methods are going to be more troublesome and are special cases and thus should be treated separately.

stephenplusplus commented:
Sounds good. :+1:

jgeewax commented:
Cool - can we create a ticket to make sure we don't forget those two methods?

ryanseys commented:
@jgeewax Done: #364

stephenplusplus commented:
This is now updated to remove makePublic() and makePrivate() as well as change `scope` -> `entity`.

ryanseys commented on lib/storage/acl.js:
```diff
@@ -67,17 +69,138 @@ var util = require('../common/util.js');
 }
 
 /**
+ * A convenience object of methods to add or delete ACL permissions from an
```
I'm assuming this is public documentation so perhaps we should clarify the role of these methods is for owners. Example:  "A convenience object of methods to add and delete **owner** ACL permissions from different entities."

ryanseys commented on lib/storage/acl.js:
```diff
@@ -67,17 +69,138 @@ var util = require('../common/util.js');
+Acl.prototype.owners = {};
+
+/**
+ * A convenience object of methods to add or delete ACL permissions from an
```
Similarily, we might wish to specify this deals with the reader role. Example: "A convenience object of methods to add and delete **reader** ACL permissions from different entities."

ryanseys commented on lib/storage/acl.js:
```diff
@@ -67,17 +69,138 @@ var util = require('../common/util.js');
+Acl.prototype.readers = {};
+
+/**
+ * A convenience object of methods to add or delete ACL permissions from an
```
And finally writer role. Example: "A convenience object of methods to add and delete **writer** ACL permissions from different entities."

ryanseys commented on lib/storage/acl.js:
```diff
@@ -328,3 +451,90 @@ Acl.prototype.makeReq_ = function(method, path, query, body, callback) {
+  var roleGroup = role.toLowerCase() + 's';
+
+  this[roleGroup] = entities.reduce(function(acc, entity) {
+    var isPrefix = entity.substr(-1) === '-';
```
It's more readable to use `entity.charAt(entity.length-1)` (and it's [faster](http://jsperf.com/splice-1-vs-charat-lenght-1-vs-index)!)

ryanseys commented:
Overall LGTM, just lightly commented on a couple things.

stephenplusplus commented on lib/storage/acl.js:
```diff
@@ -67,17 +69,138 @@ var util = require('../common/util.js');
 
 /**
+ * An object of convenience methods to add or delete owner ACL permissions for a
+ * given entity.
```
@ryanseys - the first of the three re-writes suggested.

stephenplusplus commented:
Addressed those last issues and squashed down to one commit.

Docs:
- http://stephenplusplus.github.io/gcloud-node/#/docs/master/storage/bucket
- http://stephenplusplus.github.io/gcloud-node/#/docs/master/storage/file

jgeewax commented:
For the docs, can we add a "suggested example" at the beginning of the page? something so that people don't have to scroll past the ".add()" confusion?

stephenplusplus commented:
I added some examples to the top of Storage pages: http://stephenplusplus.github.io/gcloud-node/#/docs/master/storage

Let me know if you prefer other examples to showcase.

jgeewax commented:
:+1:
